server/src/main/java/io/crate/session/RowConsumerToResultReceiver.java

```diff
@@ -80,14 +80,18 @@ private void consumeIt(BatchIterator<Row> iterator) {
                     rowCount++;
                     CompletableFuture<Void> writeFuture = resultReceiver.setNextRow(iterator.currentElement());
                     if (!resultReceiver.isWritable()) {
-                        assert writeFuture != null : "writeFuture must not be null if resultReceiver is not writable";
-                        LOGGER.trace("Suspended execution after {} rows as the receiver is not writable anymore", rowCount);
-                        activeIt = iterator;
-                        writeFuture.thenRun(() -> {
-                            LOGGER.trace("Resume execution after {} rows", rowCount);
-                            resume();
-                        });
-                        return;
+                        // The `isWritable` can become false after the row has been processed. In this case the returned
+                        // future is NULL. We must continue the execution to get a new future from the next row on which
+                        // we can add the `resume` logic to.
+                        if (writeFuture != null) {
```

Comment:
A different solution would be to always return a future which we changed to avoid allocations.
Honestly, I'd tend to to ^^ as this solution will lead to another row being buffered.

Reply:
Went with an even better approach.